### PR TITLE
Software-Project TU Ilmenau Summer Semester 2017

### PR DESCRIPTION
**Pull request by 1DIce**
# Software-Project TU Ilmenau Summer Semester 2017

## Aim of the Project

This software project aims to extend the MNE-CPP framework by adding **mesh utilities** for the visualization of EEG/MEG signals. Because computation time is supposed to be kept as low as possible, efficiency of the code is one of the main criterion.

## Added Components

Two new parts were added to the library: **GeometryInfo** and **Interpolation**.
GeometryInfo holds two features: Sensor Projecting and Surface Constrained Distance Calculation.
Interpolation provides methods for calculating an interpolation matrix and for interpolation of signal values.
In order to use these new features, a new item called **"SensorDataTreeItem"** was added within the Disp3D tree structure. Said item houses a worker that provides interpolated visual output data.

### Sensor Projecting

Because the sensors coordinates do not match any vertices of the used mesh, they must be assigned to some vertices. This algorithm looks up the best-matching vertex for each sensor.

### Surface Constrained Distance Calculation (SCDC)

Because the used meshes often are uneven, the surface constrained distance between two vertices differs from the euclidian distance. In order to avoid this, we implemented an algorithm that calculates approximate surface constrained distances by solving the shortest path problem on undirected graphs, using the adjacency information that is provided in each mesh.  

### Interpolation Matrix

The interpolation of activity levels between sensors is achieved by using an interpolation matrix. The user can choose different functions for calculating interpolation weights. Said matrix is later used for real-time interpolation of sensor signals.

### SensorDataTreeItem

To use the new interpolation features, a new item was added to the Disp3D tree structure. It allows changes to parameters of the visualization, such as signal averaging etc. Two meta item types **CancelDistance** and **InterpolationFunction** were added to allow said changes.

### RtSensorDataWorker

In order to periodically generate and publish new visual data, a worker was implemented. The latter calculates an interpolation matrix when it is initialized and then uses this matrix inside a loop to generate new visual output.

## Changed Components

In order to integrate the new features correctly, some components of Disp3D had to be changed.

### Data 3D Tree Model

We added a method for adding a SensorDataTreeItem to the Disp3D tree structure.

### Measurement Tree Item

We added a method for adding a SensorDataTreeItem to the Disp3D tree structure.

### QVariant-Wrapping in Abstract Mesh Item

Deleted an unnecessary second wrapping of vertex colors inside 'AbstractMeshTreeItem::setVertColor'.

### Connect-Methods in Subject Tree Item

Fixed implementation of 'SubjectTreeItem::connectMeasurementToBemHeadItems' (wrong casts) and 'SubjectTreeItem::connectMeasurementToSensorItems' (wrong casts and item types).

## Feedback

Please submit feedback as soon as possible to **1Dice** and  **sheinke**.

### CODE REVIEW

**Comment by 1DIce** [comment]
im working on a fix for the failed tests.

**Comment by sheinke** [comment]
This should fix all build errors under linux. Just added the dependencies to all .pro files that are using Disp3D since it needs GeometryInfo and Interpolation. Added a missing include for interpolation example, this should fix the build errors under windows.

File: MNE/disp3D/engine/model/data3Dtreemodel.cpp [hunk 1]
```diff
@@ -482,3 +483,52 @@ void Data3DTreeModel::addItemWithDescription(QStandardItem* pItemParent, QStanda
...
 
 
+
+//*************************************************************************************************************
+
+SensorDataTreeItem* Data3DTreeModel::addSensorData(const QString& sSubject,
+                                        const QString& sMeasurementSetName,
+                                        const MatrixXd& matSensorData,
+                                        const MNEBemSurface& tBemSurface,
+                                        const FiffEvoked& fiffEvoked,
```
Comment (LorenzE) [inline MNE/disp3D/engine/model/data3Dtreemodel.cpp]: please change FiffEvoked to FiffInfo

File: MNE/disp3D/engine/model/items/measurement/measurementtreeitem.h [hunk 2]
```diff
@@ -160,6 +162,26 @@ class DISP3DNEWSHARED_EXPORT MeasurementTreeItem : public AbstractTreeItem
...
+    * @param[in] sSensorType            Name of the sensor type EEG or MEG.
+    * @param[in] dCancelDist            Distances higher than this are ignored for the interpolation
+    * @param[in] sInterpolationFunction Function that computes interpolation coefficients using the distance values
+    *
+    * @return                           Returns a pointer to the added tree item. (Default would be a NULL pointer if no item was added.)
+    */
+    SensorDataTreeItem* addData(const MatrixXd& tSensorData,
+                                const MNELIB::MNEBemSurface &bemSurface,
+                                const FIFFLIB::FiffEvoked &fiffEvoked,
+                                const QString sSensorType,
```
Comment (LorenzE) [inline MNE/disp3D/engine/model/items/measurement/measurementtreeitem.h]: please pass QString as reference

File: MNE/disp3D/engine/model/items/sensordata/sensordatatreeitem.cpp [hunk 3]
```diff
@@ -0,0 +1,556 @@
...
+
+void SensorDataTreeItem::onNumberAveragesChanged(const QVariant& iNumAvr)
+{
+    if(iNumAvr.canConvert<int>()) {
+        if(m_pSensorRtDataWorker) {
+            m_pSensorRtDataWorker->setNumberAverages(iNumAvr.toInt());
+        }
+    }
+}
+
```
Comment (LorenzE) [inline MNE/disp3D/engine/model/items/sensordata/sensordatatreeitem.cpp]: please add spacers //*************************************************************************************************************

File: MNE/geometryInfo/geometryinfo.cpp [hunk 4]
```diff
@@ -0,0 +1,318 @@
+//=============================================================================================================
+/**
+* @file     geometryinfo.cpp
+* @author   Lars Debor <lars.debor@tu-ilmenau.de>;
+*           Matti Hamalainen <msh@nmr.mgh.harvard.edu>
+* @version  1.0
+* @date     Mai, 2017
```
Comment (LorenzE) [inline MNE/geometryInfo/geometryinfo.cpp]: please use english: May. Please check other files too.

File: MNE/geometryInfo/geometryinfo.h [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+* HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+* NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+* POSSIBILITY OF SUCH DAMAGE.
+*
+*
+* @brief     GeometryInfo class declaration.
+*
+*/
+
+#ifndef GEOMETRYINFO_GEOMETRYINFO_H
```
Comment (LorenzE) [inline MNE/geometryInfo/geometryinfo.h]: please change this to +#ifndef GEOMETRYINFO_H. Please check other files too

File: MNE/interpolation/interpolation.h [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+* HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+* NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+* POSSIBILITY OF SUCH DAMAGE.
+*
+*
+* @brief     Interpolation class declaration.
+*
+*/
+
+#ifndef INTERPOLATION_INTERPOLATION_H
```
Comment (LorenzE) [inline MNE/interpolation/interpolation.h]: change pls INTERPOLATION_H

**Pull request merged**